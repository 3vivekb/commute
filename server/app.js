const bodyParser = require('body-parser')
const express = require('express')
const jwt = require('express-jwt')
const morgan = require('morgan')
const path = require('path')
const html = require('@conveyal/woonerf/html')

const env = require('./utils/env').env
const routes = require('./routes')

const app = express()

// middleware
let jwtMiddleWare
if (env.AUTH0_SIGNING_CERTIFICATE && process.env.NODE_ENV !== 'test') {
  jwtMiddleWare = jwt({
    algorithms: ['HS256', 'RS256'],
    secret: env.AUTH0_SIGNING_CERTIFICATE
  })
} else {
  jwtMiddleWare = (req, res, next) => {
    req.user = {
      sub: 'test-user'
    }
    next()
  }
}
app.use(morgan('combined'))
app.use(bodyParser.json({ limit: '50mb' }))

// static assets
app.use('/assets', express.static(path.resolve(__dirname, '../assets')))

// api
routes(app, jwtMiddleWare)

const htmlString = html({
  staticHost: process.env.STATIC_HOST || '/',
  title: 'Commute'
})

// webapp
app.get('*', (req, res) => {
  res.status(200).type('html').send(htmlString)
})

module.exports = app
