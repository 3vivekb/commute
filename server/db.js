const mongoose = require('mongoose')

mongoose.Promise = global.Promise // default is mpromise...ugh

let URI = process.env.MONGODB_URI

if (!URI) {
  const baseConnectionString = 'mongodb://localhost:27017/'
  let dbName = 'commute'
  if (process.env.NODE_ENV === 'test') {
    const uuid = require('uuid')
    dbName += uuid.v4()
  }
  URI = `${baseConnectionString}${dbName}`
}

const db = module.exports = mongoose.createConnection(URI)

function dropDbIfTesting (callback) {
  if (!callback) {
    callback = () => 'no-op'
  }
  if (process.env.NODE_ENV === 'test') {
    console.log('dropping test database ' + URI)
    db.dropDatabase(callback)
  } else {
    callback()
  }
}

db.on('connected', () => console.log('Mongoose default connection open to ' + URI))
db.on('error', (err) => console.log('Mongoose default connection error: ' + err))
db.on('disconnected', () => {
  console.log('Mongoose default connection disconnected')
  dropDbIfTesting()
})

// If the Node process ends, close the Mongoose connection
process.once('SIGINT', () => {
  dropDbIfTesting(() => {
    db.close(() => {
      console.log('Mongoose default connection disconnected through app termination')
      process.exit(0)
    })
  })
})

// For NODEMON
process.once('SIGUSR2', () => {
  dropDbIfTesting(() => {
    db.close(() => {
      console.log('Mongoose default connection disconnected through nodemon restart')
      process.kill(process.pid, 'SIGUSR2')
    })
  })
})
