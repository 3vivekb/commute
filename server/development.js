const app = require('./app')
const db = require('./db')

const PORT = process.env.PORT || 9966

db.on('connected', () =>
  app.listen(PORT, (err) =>
    err ? console.error(err) : console.log('GraphQL API now running on port', PORT)))
