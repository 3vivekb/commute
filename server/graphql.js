const {buildSchema} = require('graphql')

const {
  Organization
} = require('./models')

const schema = buildSchema(`
  input Find {
    _id: ID
    owner: String
  }

  type Point {
    type: String!
    coordinates: [Float]
  }

  type Commuter {
    _id: ID!
    name: String!
    location: Point
  }

  type Group {
    _id: ID!
    name: String!
    organization: Organization!
    commuters: [Commuter]
  }

  type Organization {
    _id: ID!
    owner: String!
    name: String!
    update(name: String!): Organization
  }

  type Site {
    _id: ID!
    location: Point!
    organization: Organization!
  }

  type SiteAnalysis {
    _id: ID!
    group: Group!
    organization: Organization!
    site: Site!
  }

  type Query {
    organization(find: Find): Organization
    organizations: [Organization]
  }

  type Mutation {
    createOrganization(name: String!): Organization
    deleteOrganization(name: String!): Organization
    organization(_id: ID!): Organization
    updateOrganization(name: String! newName: String!): Organization
  }
`)

const rootValue = {
  createOrganization (properties, request) {
    const o = new Organization({
      owner: request.user || 'Trevor',
      name: properties.name
    })
    return o.save()
  },
  deleteOrganization (properties, request) {
    return Organization
      .findOne({ name: properties.name })
      .exec((res) => {
        return Organization.remove(res).exec()
      })
  },
  organization ({find}, request) {
    return Organization.findOne(find).exec()
  },
  organizations (properties, request) {
    return Organization
      .find()
      .exec()
  },
  updateOrganization (properties, request) {
    return Organization.findOneAndUpdate({
      name: properties.name
    }, {
      name: properties.newName
    }).exec()
  }
}

module.exports = {schema, rootValue}
