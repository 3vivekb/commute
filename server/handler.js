const {graphql} = require('graphql')
const {schema, root} = require('./graphql')

module.exports.graphql = (event, context, callback) =>
  graphql(schema, event.body.query, root, {}, event.body.variables)
    .then((response) => callback(null, response))
    .catch((error) => callback(error))
