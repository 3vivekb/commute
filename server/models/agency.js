const {Schema} = require('mongoose')

const db = require('../db')
const Organization = db.model('Organization', require('./organization'))
const trashPlugin = require('./plugins/trash')
const dbUtils = require('../utils/db')

const schema = new Schema({
  name: {
    required: true,
    type: String
  }
})

schema.plugin(trashPlugin)

schema.pre('save', dbUtils.makeCascadeDeleteModelsFn('agencyId', [Organization]))

module.exports = schema
