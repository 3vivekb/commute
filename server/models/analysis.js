const {Schema} = require('mongoose')

const db = require('../db')
const Commuter = db.model('Commuter', require('./commuter'))
const Group = db.model('Group', require('./group'))
const trashPlugin = require('./plugins/trash')
const Site = db.model('Site', require('./site'))
const Trip = db.model('Trip', require('./trip'))
const dbUtils = require('../utils/db')
const later = require('../utils/later')

const schema = new Schema({
  calculationStatus: {
    default: 'calculating',
    required: true,
    type: String
  },
  groupId: {
    ref: 'Group',
    required: true,
    type: Schema.Types.ObjectId
  },
  name: {
    required: true,
    type: String
  },
  numCommuters: {
    default: 0,
    required: true,
    type: Number
  },
  organizationId: {
    ref: 'Organization',
    required: true,
    type: Schema.Types.ObjectId
  },
  siteId: {
    ref: 'Site',
    required: true,
    type: Schema.Types.ObjectId
  },
  summary: {
    // fields
    avgTravelTime: Number,
    avgDistance: Number,
    savingsPerTrip: Number,
    savingsPerTripYear: Number,
    savingsTotalPerDay: Number,
    savingsTotalPerYear: Number,

    // mongoose
    default: {
      avgTravelTime: 0,
      avgDistance: 0,
      savingsPerTrip: 0,
      savingsPerTripYear: 0,
      savingsTotalPerDay: 0,
      savingsTotalPerYear: 0
    },
    required: true,
    type: Schema.Types.Mixed
  }
})

schema.plugin(trashPlugin)

schema.pre('save', dbUtils.makeCascadeDeleteModelsFn('analysisId', [Trip]))

schema.pre('save', true, function (next, done) {
  next()

  const self = this

  if (['skipCalculation', 'calculated'].indexOf(this.calculationStatus) !== -1 ||
    this.trashed) return done()

  const profiler = require('../utils/profiler')

  // initiate profile analysis of all commuters in group
  // find all commuters via groupId
  // make sure group isn't trashed
  const ensureGroupExists = Group.count({ _id: this.groupId, trashed: undefined }).exec()
  const findCommuters = Commuter.find({ groupId: this.groupId, trashed: undefined }).exec()
  const getSite = Site.findOne({ _id: this.siteId, trashed: undefined }).exec()
  Promise.all([findCommuters, getSite, ensureGroupExists])
    .then((results) => {
      // verify that group, commuters and site exist
      const commuters = results[0]
      const site = results[1]
      const groupCount = results[2]

      if (commuters.length === 0) return done('No commuters found!')
      if (!site) return done('Site not found!')
      if (groupCount === 0) return done('Group not found!')

      self.numCommuters = commuters.length

      later(() => {
        profiler({
          analysisId: self._id,
          commuters,
          site
        })
      })
      done()
    })
    .catch((err) => {
      console.error(err)
      self.calculationStatus = 'error'
      done()
    })
})

module.exports = schema
