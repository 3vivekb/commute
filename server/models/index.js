const db = require('../db')

module.exports = {
  Commuter: db.model('Commuter', require('./commuter')),
  MultiSite: db.model('MultiSite', require('./multi-site')),
  Site: db.model('Site', require('./site'))
}
