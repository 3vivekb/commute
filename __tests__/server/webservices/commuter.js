/* globals afterAll, describe, expect */

import mongoose from 'mongoose'

import {Commuter} from '../../../server/models'

import {makeRestEndpointTests} from '../../test-utils/server'

describe('commuter', () => {
  afterAll(() => {
    mongoose.disconnect() // disconnect from mongo to end running of tests
  })

  const initCommuterData = {
    address: '123 Main St',
    coordinate: {
      lat: 12,
      lon: 34
    },
    name: 'test-commuter',
    groupId: mongoose.Types.ObjectId()
  }

  makeRestEndpointTests({
    endpoints: {
      'Collection GET': {},
      'Collection POST': {
        creationData: initCommuterData,
        customAssertions: (json) => {
          expect(json[0].name).toBe('test-commuter')
        }
      },
      'DELETE': {
        initData: initCommuterData
      },
      'GET': {
        initData: initCommuterData
      },
      'PUT': {
        customAssertions: (modelData, json) => {
          expect(modelData.name).toBe('updated name')
          expect(json.name).toBe('updated name')
        },
        initData: initCommuterData,
        updateData: {
          name: 'updated name'
        }
      }
    },
    foreignKeys: ['groupId'],
    geocodePlugin: true,
    model: Commuter,
    name: 'commuter'
  })
})
