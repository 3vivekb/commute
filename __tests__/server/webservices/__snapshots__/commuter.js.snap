exports[`commuter rest endpoint should add model 1`] = `
Array [
  Object {
    "__v": 0,
    "address": "123 Main St",
    "coordinate": Object {
      "lat": 12,
      "lon": 34,
    },
    "geocodeConfidence": -1,
    "name": "test-commuter",
  },
]
`;

exports[`commuter rest endpoint should get model 1`] = `
Object {
  "__v": 0,
  "address": "123 Main St",
  "coordinate": Object {
    "lat": 12,
    "lon": 34,
  },
  "geocodeConfidence": -1,
  "name": "test-commuter",
}
`;

exports[`commuter rest endpoint should update model 1`] = `
Object {
  "__v": 0,
  "address": "123 Main St",
  "coordinate": Object {
    "lat": 12,
    "lon": 34,
  },
  "geocodeConfidence": -1,
  "name": "updated name",
}
`;
